plugins/catalog-backend-module-gitlab/src/lib/client.ts

```diff
@@ -107,21 +115,113 @@ export class GitLabClient {
+                    nodes {
                       id
+                      name
+                      description
+                      fullPath
+                      parent {
+                        id
+                      }
+                    }
+                    pageInfo {
+                      endCursor
+                      hasNextPage
                     }
                   }
-                  pageInfo {
-                    endCursor
-                    hasNextPage
+                }
+              }
+            `,
+          }),
+        },
+      ).then(r => r.json());
+      if (response.errors) {
+        throw new Error(`GraphQL errors: ${JSON.stringify(response.errors)}`);
+      }
+
+      if (!response.data.group?.descendantGroups?.nodes) {
+        this.logger.warn(
+          `Couldn't get groups under ${groupPath}. The provided token might not have sufficient permissions`,
+        );
+        continue;
+      }
+
+      const groupsData = response.data.group.descendantGroups.nodes;
+
+      for (let i = 0; i < groupsData.length; i++) {
+        const groupItem = groupsData[i];
+
+        const formattedGroupResponse = {
+          id: Number(groupItem.id.replace(/^gid:\/\/gitlab\/Group\//, '')),
+          name: groupItem.name,
+          description: groupItem.description,
+          full_path: groupItem.fullPath,
+          parent_id: Number(
+            groupItem.parent.id.replace(/^gid:\/\/gitlab\/Group\//, ''),
+          ),
+        };
+
+        items.push(formattedGroupResponse);
+      }
+      ({ hasNextPage, endCursor } =
+        response.data.group.descendantGroups.pageInfo);
+    } while (hasNextPage);
+    return { items };
+  }
+
+  async getGroupMembers(
+    groupPath: string,
+    relations: string[],
+  ): Promise<PagedResponse<GitLabUser>> {
+    const items: GitLabUser[] = [];
+    let hasNextPage: boolean = false;
+    let endCursor: string | null = null;
+    do {
+      const response: GitLabGroupMembersResponse = await fetch(
+        `${this.config.baseUrl}/api/graphql`,
+        {
+          method: 'POST',
+          headers: {
+            ...getGitLabRequestOptions(this.config).headers,
+            ['Content-Type']: 'application/json',
+          },
+          body: JSON.stringify({
+            variables: { group: groupPath, relations: relations, endCursor },
+            query: /* GraphQL */ `
+              query getGroupMembers(
+                $group: ID!
+                $relations: [GroupMemberRelation!]
+                $endCursor: String
+              ) {
+                group(fullPath: $group) {
+                  groupMembers(
+                    first: 100
+                    relations: $relations
+                    after: $endCursor
+                  ) {
+                    nodes {
+                      user {
+                        id
+                        username
+                        publicEmail
```

Comment:
I am still not 100% confident about the field used for email. Did you had the chance to cross-check with GitLab?

Reply:
I have opened a Support case with GitLab to clarify. At this time, we don' see any email fields auto-populated for GitLab.com user accounts (unless the user take a manual action) and we feel like this is potentially a misconfiguration on our side of the SAML SSO setup (TBC in the GitLab Support case).

In our case, we're not really worried about users not having emails populated on the `User` entity. We just want the `User` entity ingested so we can enable the sign-in resolver and lookup the user in the catalog by username instead.

Reply:
So after some back and forth with GitLab.com support, here's the gist of how things work:

> GraphQL: 
 
> If you are using the [GraphQL endpoint](https://docs.gitlab.com/ee/api/graphql/users_example.html) to query users, you will only see the fields publicEmail and commitEmail populated if they have been explicitly set.
 
> The field publicEmail is not set by default, which means that you will only get a value returned here if you actively change its value.
 
> The value for a user's commit email defaults to the primary email address. However, GraphQL will not show this. It will only return a value for commitEmail if you actively change the value from the default value. If you don't actively change this, GraphQL won't return a value.

With that, switching to `commitEmail` seems like a more sane default. Thanks for querying the logic here @matteosilv 👍

Reply:
You forgot to change this to `commitEmail` right? But still try to read a `commitEmail` field below